src/main/java/com/redhat/devtools/lsp4ij/client/features/EditorBehaviorFeature.java

```diff
@@ -176,9 +174,49 @@ public static boolean enableTextMateNestedBracesImprovements(@NotNull PsiFile fi
      * @return true if the semantic tokens-based file view provider is enabled; otherwise false
      */
     public static boolean enableSemanticTokensFileViewProvider(@NotNull PsiFile file) {
-        return LanguageServiceAccessor.getInstance(file.getProject()).hasAny(
-                file.getVirtualFile(),
-                ls -> ls.getClientFeatures().getEditorBehaviorFeature().isEnableSemanticTokensFileViewProvider(file)
+        return isEnabled(file, EditorBehaviorFeature::isEnableSemanticTokensFileViewProvider);
+    }
+
+    private interface FeatureFlagChecker {
+        boolean isEnabled(@NotNull EditorBehaviorFeature editorBehaviorFeature, @NotNull PsiFile file);
+    }
+
+    private static boolean isEnabled(
+            @NotNull PsiFile file,
+            @NotNull FeatureFlagChecker featureFlagChecker
+    ) {
+        return isEnabled(file.getProject(), file.getVirtualFile(), featureFlagChecker);
+    }
+
+    private static boolean isEnabled(
+            @NotNull Project project,
+            @Nullable VirtualFile virtualFile,
+            @NotNull FeatureFlagChecker featureFlagChecker
+    ) {
+        if (virtualFile == null) return false;
+
+        // Create a compact cache key if possible
+        String virtualFileId = virtualFile instanceof VirtualFileWithId virtualFileWithId ? String.valueOf(virtualFileWithId.getId()) : virtualFile.getPath();
+        Key<CachedValue<Boolean>> cacheKey = Key.create(featureFlagChecker.getClass().getName() + "::" + virtualFileId);
+        return CachedValuesManager.getManager(project).getCachedValue(
+                project,
+                cacheKey,
+                () -> {
+                    PsiFile file = LSPIJUtils.getPsiFile(virtualFile, project);
```

Comment:
As mentioned above, we must re-establish the PSI file here inside of the caching logic to avoid having the original PSI file unintentionally included as a cache dependency.